### DISCUSSION TITLE
'repeatable' & 'subfields' & 'store_in' issue

### DISCUSSION THREAD

**Discussion by blondie63** [Laravel-Backpack/community-forum#957] [Q&A (Help)] [answered]
I'm using this way https://backpackforlaravel.com/docs/6.x/crud-fields#optional-fake-field-attributes-stores-fake-attributes-as-json-in to store json data to a repeatable field.

This my CRUD field:
```
        CRUD::Field([
            'name' => 'giocate_superenalotto',
            'label' => 'Giocate SuperEnalotto',
            'type' => 'repeatable',
            'subfields' => [
                [
                    'name'  => 'giocata',
                    'label' => "Giocata",
                    'type'  => 'custom_superenalotto_numbers',
                ],
            ],
            'wrapperAttributes' => [
                'class' => 'form-group giocate_superenalotto col-md-12'
            ],
            'fake'     => true,
            'store_in' => 'giocate'
        ]);
```
Store method works fine and i see data on db fields
But when use update method i can't load data on subfields

Putting a dd(get_defined_vars()); on top of my custom_superenalotto_numbers.blade.php i see this:
<img width="667" alt="bkp6-store_in" src="https://github.com/Laravel-Backpack/community-forum/assets/1261851/315290d6-d0ba-49c8-bdcc-f59c3671e080">

I don't understand why the custom field blade works fine creating and not updating :(

Some help ?

**Comment by jcastroa87** [comment]
Hello @blondie63 can you share the content of custom_superenalotto_numbers field.

Thanks.

**Reply by blondie63** [reply]
Sure @jcastroa87 this my custom_superenalotto_numbers.blade.php
```
@php
    $giocataData = $field['value'] ?? [];
@endphp
<div class="row">
    @for ($i = 1; $i <= 6; $i++)
        <div class="col-1">
            <div class="form-group">
                <label for="n{{ $i }}" class="small">N{{ $i }}</label>
                <input type="number" id="n{{ $i }}" name="{{ $field['name'] }}[n{{ $i }}]" class="form-control lottery-number" value="{{ old('n' . $i, $giocateData['n' . $i] ?? '') }}" min="1" max="90">
            </div>
        </div>
    @endfor
    <div class="col-1">
        <div class="form-group">
            <label for="superstar" class="small">Superstar</label>
            <input type="number" id="superstar" name="{{ $field['name'] }}[superstar]" class="form-control lottery-number" value="{{ old('superstar', $giocateData['superstar'] ?? '') }}" min="1" max="90">
        </div>
    </div>
</div>
```

**Reply by blondie63** [reply]
Hi @jcastroa87 Have you found the issue ?

**Reply by jcastroa87** [reply]
hello @blondie63 i check the issue and repeatable is working ok, the problem is in the custom field.

Dont forget in your model set fillable, fakeColumns and casts

```
    protected $fillable = ['name', 'description', 'giocate'];
    // protected $hidden = [];
    protected $fakeColumns = ['giocate'];
    protected $casts = [
        'giocate' => 'array',
    ];
```

In repeatable is not needed set fake and store_in, but this is not the real problem.

The problem is you are trying to save many fields inside your custom field, you will need to change this to work similar to date_range, check that's field code and declaration in controller.

Like and example this work perfect.
```
CRUD::field('name')->type('text');
        CRUD::field('description')->type('textarea');
        CRUD::field([
            'name' => 'giocate',
            'label' => 'Giocate SuperEnalotto',
            'type' => 'repeatable',
            'subfields' => [
                [
                    'name'  => 'start_date,end_date',
                    'label' => 'Giocata',
                    'type'  => 'date_range',
                ],
            ],
            'wrapperAttributes' => [
                'class' => 'form-group giocate_superenalotto col-md-12'
            ],
        ]);
```

Cheers.

**Reply by blondie63** [reply]
Hi @jcastroa87 , thanks for your collaboration..
I've done some tests after added protected $fakeColumns = ['giocate'];
and i've seen that without 'fake'     => true, data are not saved in 'giocate' field
i need to use 'store_in' => 'giocate' because i've two similar Crud fields:

```
        CRUD::Field([
            'name' => 'giocate_superenalotto',
            'label' => 'Giocate SuperEnalotto',
            'type' => 'repeatable',
            'subfields' => [
                [
                    'name'  => 'giocata',
                    'label' => "Giocata",
                    'type'  => 'custom_superenalotto_numbers',
                ],
            ],
            'wrapperAttributes' => [
                'class' => 'form-group giocate_superenalotto col-md-12'
            ],
            'fake'     => true,
            'store_in' => 'giocate'
        ]);
        CRUD::Field([
            'name' => 'giocate_eurojackpot',
            'label' => 'Giocate Eurojackpot',
            'type' => 'repeatable',
            'subfields' => [
                [
                    'name'  => 'giocata',
                    'label' => "Giocata",
                    'type'  => 'custom_eurojackpot_numbers',
                ],
            ],
            'wrapperAttributes' => [
                'class' => 'form-group giocate_eurojackpot col-md-12'
            ],
            'fake'     => true,
            'store_in' => 'giocate'
        ]);
```
Data is always saved on db table:
<img width="1105" alt="bkp6-store-in-t2" src="https://github.com/Laravel-Backpack/community-forum/assets/1261851/f40688db-464e-4d31-85ff-905776b08697">

but 'giocate' are not loaded from blade also if i seen them, look:
<img width="681" alt="bkp6-repe-t2" src="https://github.com/Laravel-Backpack/community-forum/assets/1261851/ed302b47-78e2-44a3-9723-589203061d1c">

I don't understand why the blade works fine to save but not to load..
Do you understand ?

**Comment by jcastroa87** [comment]
Hi @blondie63 as i named before you need to take as an example "date_range" there is saving multiple values for a field and is you check in detail, declaration is an array, so check thats field in detail and you will be able to save and then recovery many values in a field, is needed to handle manually many things in this case.

Cheers.

**Reply by blondie63** [reply]
@jcastroa87 Sorry but I don't understand why as an example a different type for the subfields..
My custom type correctly saves N lines in the 'giocate' json field
I just have to figure out how to get them back..
Looking at the result of dd(get_defined_vars()); 
the code $giocataData = $field['value'] should return the array I indicated in the screenshot but instead it returns null!
I don't understand this :(

**Reply by jcastroa87** [reply, marked as answer]
your custom field is not correctly handle the saving, the name of subfield need to be consistent with the name of saving values, so per example, the name of your field is "giocata", tha save json is "[{"n1":"1","n2":"2","n3":"3","n4":"4","n5":"5","n6":"6","superstar":"7"}]", there is no way Backpack will know nX is the same as giocata, if you check date_range, the name of field is 'start_date,end_date', the save json is "[{"start_date":"2024-04-16 00:00:00","end_date":"2024-05-17 23:59:59"}]", you see the point?

Cheers.

**Reply by pxpm** [reply]
@blondie63 the value is in `giocate_supernalotto` repeatable field, not in your subfields. Subfields then get their values from the "parent" field.
There is only one subfield, and none of those array keys map to that field, that's why you get null instead of the expected value. 

From Backpack perspective your repeatable only has 1 subfield called `giocata`. 
What Backpack tries to do is `get from the parent repeatable values the value for giocata subfield`. There is no such field in the saved array. 

Your array in the database should be something like:
```php
[
    0 => [ // row number 1
        'giocata' => [
            'n1' => 1,
            'n2' => 2
        ]
    ],
    1 => [ // row number 2
        'giocata' => [
            'n1' => 3,
            'n2' => 4
        ]
    ],
]
```

If you had another subfield it should be:
```php
[
    0 => [ // row number 1
        'giocata' => [
            'n1' => 1,
            'n2' => 2
        ],
        'other_subfield' => 'some value',
    ],
 ]
```

I would advise you like @jcastroa87 said to have a look at the date_range, because you will need to do something similar to process the "raw value" and assign the bits that belong to each field. 
![image](https://github.com/Laravel-Backpack/community-forum/assets/7188159/2b18dc16-39f4-40c9-b3b5-e32319774760)

Let me know if I made myself clear. 

Cheers

**Comment by blondie63** [comment]
I tried in every way to manage an array inside the "game" field but I couldn't!
In the end I opted for a text field that contains the numbers separated by commas (for example "2,5,12,5,78") and then I split the numbers and manage loading and saving separately via the store() and update( methods ) and display blade.
This is how it works even if I find it inelegant :)
